astrobee/behaviors/inspection/src/camera_projection.cc

```diff
@@ -0,0 +1,444 @@
+        ROS_DEBUG_STREAM("p_near 4 " << p_near(0, 3) / p_near(3, 3) << " " << p_near(1, 3) / p_near(3, 3) << " "
+                                     << p_near(2, 3) / p_near(3, 3));
+
+        Eigen::Matrix4d corners_far;
+        corners_far << 1, -1, 1, -1, -1, -1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1;
+        Eigen::Matrix4d p_far = P_.inverse() * corners_far;
+        ROS_DEBUG_STREAM("p_far 1 " << p_far(0, 0) / p_far(3, 0) << " " << p_far(1, 0) / p_far(3, 0) << " "
+                                    << p_far(2, 0) / p_far(3, 0));
+        ROS_DEBUG_STREAM("p_far 2 " << p_far(0, 1) / p_far(3, 1) << " " << p_far(1, 1) / p_far(3, 1) << " "
+                                    << p_far(2, 1) / p_far(3, 1));
+        ROS_DEBUG_STREAM("p_far 3 " << p_far(0, 2) / p_far(3, 2) << " " << p_far(1, 2) / p_far(3, 2) << " "
+                                    << p_far(2, 2) / p_far(3, 2));
+        ROS_DEBUG_STREAM("p_far 4 " << p_far(0, 3) / p_far(3, 3) << " " << p_far(1, 3) / p_far(3, 3) << " "
+                                    << p_far(2, 3) / p_far(3, 3));
+      }
+      // Eliminate point
+      return false;
+    }
+
+
+    return true;
+  }
+
+
+  bool CameraView::GetCamXYFromPoint(const geometry_msgs::Point point, int& x, int& y) {
+    // Initialize x,y
+    x = 0; y = 0;
+    // Get current camera position if it's not given
+    geometry_msgs::TransformStamped robot_pose;
+    try {
+      robot_pose = tf_buffer_.lookupTransform("world", "body", ros::Time(0), ros::Duration(1.0));
+    } catch (tf2::TransformException &ex) {
+      ROS_ERROR("Failed getting transform: %s", ex.what());
+      return false;
+    }
+    return GetCamXYFromPoint(msg_conversions::ros_transform_to_ros_pose(robot_pose.transform), point, x, y);
+  }
+
+  // Get point from camera pixel location and point cloud
+  bool CameraView::GetPointFromXYD(const sensor_msgs::PointCloud2 pCloud, const int x, const int y,
+                                   geometry_msgs::Point& point) {
+    // Convert from u (column / width), v (row/height) to position in array
+    // where X,Y,Z data starts
+    int arrayPosition = x*pCloud.row_step + y*pCloud.point_step;
+
+    // compute position in array where x,y,z data start
+    int arrayPosX = arrayPosition + pCloud.fields[0].offset;  // X has an offset of 0
+    int arrayPosY = arrayPosition + pCloud.fields[1].offset;  // Y has an offset of 4
+    int arrayPosZ = arrayPosition + pCloud.fields[2].offset;  // Z has an offset of 8
+
+    float X = 0.0;
+    float Y = 0.0;
+    float Z = 0.0;
+
+    memcpy(&X, &pCloud.data[arrayPosX], sizeof(float));
+    memcpy(&Y, &pCloud.data[arrayPosY], sizeof(float));
+    memcpy(&Z, &pCloud.data[arrayPosZ], sizeof(float));
+
+    // put data into the point p
+    point.x = X;
+    point.y = Y;
+    point.z = Z;
+
+    return true;
+  }
+  bool CameraView::InsideTarget(std::vector<int> vert_x, std::vector<int> vert_y, int test_x, int test_y) {
```

Comment:
Yes this is a well known implementation of this functionality, I will add further documentation.

Since it's only used in the CameraView context I added it to the class for easy management, if there is a need to use it outside the CameraView context we can def make it independent (it's not fully independent since the number of vertices are hardcoded to be 4).

Given the uncertainty surrounding the definition what is defined as a target (target position and size), the differences between results from using floating point or integers here should be irrelevant. I rounded to integers due to a natural connection between the value and the pixel, but I do not have any preference since I didn't think it mattered in a practical way.